runtime/vm/OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler.cpp

```diff
@@ -216,6 +225,70 @@ OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_initCifNativeTh
 	ffiTypeHelpers.freeStructFFIType(returnType);
 	goto done;
 }
+
+#if JAVA_SPEC_VERSION >= 22
+/* openj9.internal.foreign.abi.InternalDowncallHandler: private static native boolean isFfiProtoEnabled(); */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_isFfiProtoEnabled(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	bool isFfiProtoOn = (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_FFI_PROTO)) ? true : false;
+
+	VM_OutOfLineINL_Helpers::returnSingle(currentThread, isFfiProtoOn, 0);
+	return rc;
+}
+
+/**
+ * openj9.internal.foreign.abi.InternalDowncallHandler: private native void resolveBoundHandle();
+ *
+ * @brief Resolve the bounded handle via MethodHandleResolver.ffiCallLinkCallerMethod() in advance
+ *        when calling into the interpreter via MH.linkToNative()
+ *
+ * @return void
+ */
+VM_BytecodeAction
+OutOfLineINL_openj9_internal_foreign_abi_InternalDowncallHandler_resolveBoundHandle(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	j9object_t invokeCache = NULL;
+
+	jobject nativeInvoker = j9jni_createGlobalRef((JNIEnv *)currentThread, J9_JNI_UNWRAP_REFERENCE(currentThread->sp), JNI_FALSE);
+	if (NULL == nativeInvoker) {
+		rc = GOTO_THROW_CURRENT_EXCEPTION;
+		setNativeOutOfMemoryError(currentThread, 0, 0);
+		goto done;
+	}
+
+	/* The resolution is performed by MethodHandleResolver.ffiCallLinkCallerMethod()
+	 * to obtain a MemberName object plus appendix in a two element object array.
+	 */
+	VM_OutOfLineINL_Helpers::buildInternalNativeStackFrame(currentThread, method);
+	invokeCache = resolveFfiCallInvokeHandle(
+					currentThread,
+					J9VMOPENJ9INTERNALFOREIGNABIINTERNALDOWNCALLHANDLER_BOUNDMH(
+									currentThread,
+									J9_JNI_UNWRAP_REFERENCE(nativeInvoker)));
```

Comment:
Please indent just two tabs per level (not four).

Reply:
Fixed.